sys/kern/devfs.c

```diff
@@ -80,8 +81,24 @@ static int devfs_add_entry(devfs_node_t *parent, const char *name,
   return 0;
 }
 
+int devfs_remove(devfs_node_t *dn) {
+  SCOPED_MTX_LOCK(&devfs.lock);
+
+  assert(dn->dn_parent != NULL);
+
+  /* Only allow removal of empty directories. */
+  if (dn->dn_vnode->v_type == V_DIR && !TAILQ_EMPTY(&dn->dn_children))
+    return EINVAL;
+
+  TAILQ_REMOVE(&dn->dn_parent->dn_children, dn, dn_link);
+  vnode_drop(dn->dn_vnode);
+  kfree(M_STR, dn->dn_name);
+  kfree(M_DEVFS, dn);
```

Comment:
Shouldn't we wait with freeing memory until the reference counter of `vnode` hits 0 (or equally, until `VOP_RECLAIM` is called)? That dev file can possibly be opened in some processes.

Reply:
Yes, the file could still be opened, but the devfs node is not accessible from the vnode (at least not directly), so we're good as long as the device driver doesn't access the devfs node, which it shouldn't, since it presumably called `devfs_remove()` in the first place.

Reply:
The goal of `devfs_remove` is to remove the device from the filesystem, so that it's no longer visible. The device driver implements `VOP_RECLAIM()` and frees its data structures there.

Reply:
I'm still concerned about that situation:
1. a userspace program opens a device
2. then the device is removed by the kernel
3. user does `fstat(fd, ..)` on that opened device
4. fstat accesses the fields in devfs node, but it is already freed

Reply:
I think I see what you mean: the devfs node has some attributes (e.g. owner's uid and gid) that need to be accessible as long as the file descriptor is open. You're right: we can't free the devfs node until the vnode's refcount drops to 0 - we need to do it in `VOP_RECLAIM()`.

Reply:
...which poses some problems, as device drivers overwrite the vnode's vnodeops with their own implementation... so every device driver that supports device removal must implement `VOP_RECLAIM()` and free the devfs node there.

Reply:
Ok, I have now separated unlinking the devfs node from the tree and freeing it.